File: src/python/graph/bfs/traversal.py
from collections import deque


def traverse(graph, root):
    if root not in graph:
        return None

    explored = set()
    result = []
    queue = deque([root])

    while queue:
        vertex = queue.popleft()
        if vertex not in explored:
            explored.add(vertex)
            result.append(vertex)
            queue += graph.get(vertex) or []

    return result


def main():
    graph = {
        # first subgraph
        "A": ["B", "C", "D"],
        "B": ["E"],
        "C": ["F", "G"],
        "F": ["H", "I"],

        # second subgraph
        "J": ["K", "L", "M"],
        "L": ["N", "O"]
    }

    result = traverse(graph, "A")

    print(result) # ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I']


if __name__ == "__main__":
    main()

File: src/swift/graph/bfs/Traversal.swift
typealias Graph<T: Hashable> = [T: [T]]


func traverse<T>(in graph: Graph<T>, withRoot root: T) -> [T]? {
    guard graph[root] != nil else { return nil }

    var explored = Set<T>()
    var queue = [root]
    var popIndex = 0 // points to the most left item in queue

    while popIndex < queue.count {
        let vertex = queue[popIndex]
        popIndex += 1

        if !explored.contains(vertex) {
            explored.insert(vertex)
            if let successors = graph[vertex] {
                queue += successors
            }
        }
    }
    return queue
}


func main() {
    let graph = [
        // first subgraph
        "A": ["B", "C", "D"],
        "B": ["E"],
        "C": ["F", "G"],
        "F": ["H", "I"],

        // second subgraph
        "J": ["K", "L", "M"],
        "L": ["N", "O"],       
    ]

    if let result = traverse(in: graph, withRoot: "A") {
        print(result) // ["A", "B", "C", "D", "E", "F", "G", "H", "I"]
    }
}